python/cudf_polars/docs/overview.md

```diff
@@ -144,11 +144,69 @@ changes. We can therefore attempt to detect the IR version
 appropriately. This should be done during IR translation in
 `translate.py`.
 
-## Adding a handler for a new plan node
+# IR design
+
+As noted, we translate the polars DSL into our own IR. This is both so
+that we can smooth out minor version differences (advertised by
+`NodeTraverser` version changes) within `cudf-polars`, and so that we
+have the freedom to introduce new IR nodes and rewrite rules as might
+be appropriate for GPU execution.
+
+To that end, we provide facilities for definition of nodes as well as
+writing traversals and rewrite rules. The abstract base class `Node`
+in `dsl/nodebase.py` defines the interface for implementing new nodes,
+and provides many useful default methods. See also the docstrings of
+the `Node` class.
+
+> ![NOTE] This generic implementation relies on nodes being treated as
+> *immutable*. Do not implement in-place modification of nodes, bad
+> things will happen.
+
+## Defining nodes
+
+A concrete node type (`cudf-polars` has expression nodes, `Expr`;
+and plan nodes, `IR`), should inherit from `Node`. Nodes have
+two types of data:
+
+1. `children`: a tuple (possibly empty) of concrete nodes;
+2. non-child: arbitrary data attached to the node that is _not_ a
+   concrete node.
+
+The base `Node` class requires that one advertise the _names_ of the
+non-child attributes in the `_non_child` class variable. The
+constructor of the concrete node should take its arguments in the
+order `*_non_child` (ordered as the class variable does) and then
+`*children`. For example, the `Sort` node, which sorts a column
+generated by an expression, has this definition:
+
+```python
+class Expr(Node):
+    children: tuple[Expr, ...]
+
+class Sort(Expr):
+    _non_child = ("dtype", "options")
+    children: tuple[Expr]
+    def __init__(self, dtype, options, column: Expr):
+        self.dtype = dtype
+        self.options = options
+        self.children = (column,)
```

Comment:
IIUC this isn't enforced by the Node class directly, but rather implicitly in that code will fail to run if you fail to follow this requirement. It probably isn't worth the effort, but in principle we could probably get close to enforcement using a metaclass to define `__init__`.

Reply:
As noted below, I would like to make construction/reconstruction of this classes reasonably fast. If the metaclass checks things at class instantiation time, that seems no better than just failing at attribute access time. If we're using it to construct the class object, we could go back and do things with dataclasses again, but if not I'd like a way to do metaclass magic that provides good type checking/LSP integration.

See also some previous discussion: https://github.com/rapidsai/cudf/pull/17016#discussion_r1792388545

Reply:
Got it, yeah then definitely not worth.